### PR TITLE
Respect "disabled" in "fillIn" and "typeIn"

### PR DESCRIPTION
**Pull request by ro0gr**
fixes: https://github.com/emberjs/ember-test-helpers/issues/680
continues: https://github.com/emberjs/ember-test-helpers/pull/681

With this PR `fillIn` and `typeIn` helpers start to fail on attempt to interact with `disabled` form controls.

There is a follow up PR, doing a similar thing for readonly(#779). Extracted it to make review easier.

---

my original intention was also to cover `click` and `doubleclick` and the rest of interaction helpers with the same logic, but seems they intentionally just skip focus for non-focusable elements:

https://github.com/emberjs/ember-test-helpers/blob/f053fe0f6ac1242cb07303a4c69600e223ed5840/addon-test-support/%40ember/test-helpers/dom/click.ts#L18-L20

Assuming changing of this behavior, might be considered a breaking change, I've extracted this part as a separate follow-up PR(#778)

### CODE REVIEW

**Comment by ro0gr** [comment]
not sure what does fail on CI, it just says:

> The command "yarn test" exited with 139.

w/o any context.

**Comment by ro0gr** [comment]
hm.. turns out, in `click` we don't even try to do anything with disabled inputs:

https://github.com/emberjs/ember-test-helpers/blob/7d8a9a507e4728122e67a4c6a15282436fe86341/addon-test-support/%40ember/test-helpers/dom/click.ts#L82-L86

I think it should throw. Will update.

**Comment by rwjblue** [comment]
> I think it should throw. Will update.

👍

**Comment by ro0gr** [comment]
@rwjblue thanks for the feedback and sorry for letting it hang, going to come back to it this week.

**Comment by rwjblue** [comment]
No problem, thanks for working on it!

**Comment by ro0gr** [comment]
Just added an error for `click`, and made sure all the other interactions fail if disabled element. rebased and squashed.

But a slight doubt appeared on my mind. I suspect it might be considered a breaking change, since we didn't fail on clicking disableds before.

I'm pretty sure there are some tests in the wild, like:

```js
this.doSmth = () => {
  this.called = true
}

await render(hbs`<Something 
  @disabled={{true}}
  @onClick={{this.doSmth}}
/>`

await click('.Something');

assert.equal(called, false)
``` 

which I believe is totally valid with existing click implementation, but would fail with this proposed change. From the other hand, not throwing for disableds in `{fill,type}-in` should not be allowed, even it'd break some tests (imo).

@rwjblue please let me know, what do you think? Should I split the PR into 2, in order to merge `{fill,type}-in` stuff first, and let the the rest wait for the next major, or is it fine for you to go as is?

**Comment by ro0gr** [comment]
> Should I split the PR into 2, in order to merge {fill,type}-in stuff first, and let the the rest wait for the next major, or is it fine for you to go as is?

extracted everything unrelated to `fillIn` and `typeIn` into another PR https://github.com/emberjs/ember-test-helpers/pull/778

File: addon-test-support/@ember/test-helpers/dom/-is-focusable.ts [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
 
 const FOCUSABLE_TAGS = ['A'];
 
 type FocusableElement = HTMLAnchorElement;
 
 // eslint-disable-next-line require-jsdoc
-function isFocusableElement(
-  element: HTMLElement | SVGElement | Element
-): element is FocusableElement {
+function isFocusableElement(element: Element): element is FocusableElement {
```
Comment (Turbo87) [inline addon-test-support/@ember/test-helpers/dom/-is-focusable.ts:9]: why did you change the types here?

File: tests/unit/dom/fill-in-test.js [hunk 2]
```diff
@@ -44,6 +44,17 @@ module('DOM Helper: fillIn', function(hooks) {
...
   });
 
+  test('filling in a disabled element', async function(assert) {
+    element = buildInstrumentedElement('input');
+    element.setAttribute('disabled', true);
+
+    await setupContext(context);
+    assert.rejects(
+      fillIn(`#${element.id}`, 'foo'),
+      new Error('Can not `fillIn` disabled [object HTMLInputElement]')
```
Comment (Turbo87) [inline tests/unit/dom/fill-in-test.js:54]: why do we have `[object HTMLInputElement]` in here?

File: addon-test-support/@ember/test-helpers/dom/-is-focusable.ts (hunk 1 above)
Comment (ro0gr) [reply addon-test-support/@ember/test-helpers/dom/-is-focusable.ts:9]: I believe this was my attempt to simlify a bit. Both `HTMLElement` are `SVGElement` derrive from `Element`, so it does seem redundant to specify each.
Also, implementation wise it just checks if it's not an `A` tag!

File: tests/unit/dom/fill-in-test.js (hunk 2 above)
Comment (ro0gr) [reply tests/unit/dom/fill-in-test.js:54]: we are passing an element instace, which is not focusable, into throw statement. Should prbably change it to use passed specifier(selector or Element) instead. Will update, thanks!

File: tests/unit/dom/fill-in-test.js [hunk 3]
```diff
@@ -44,6 +44,26 @@ module('DOM Helper: fillIn', function(hooks) {
...
+
+    assert.rejects(
+      fillIn(`[data-test-disabled]`, 'foo'),
+      new Error("Can not `fillIn` disabled '[data-test-disabled]'."),
+      'renders user selector'
+    );
+
+    assert.rejects(
+      fillIn(element, 'foo'),
+      new Error("Can not `fillIn` disabled '[object HTMLInputElement]'."),
```
Comment (ro0gr) [inline tests/unit/dom/fill-in-test.js:62]: @Turbo87 just improved error messages. See this one, and 1 assert above. Please let me know if it looks good for you.

**Comment by ro0gr** [comment]
CI is failing due to some conflicts in ember-data/ember types. I suspect this was caused by this PR https://github.com/emberjs/ember-test-helpers/pull/784, which doesn't have travis results reported in the PR for some reason, so we've missed the breakage. 

I think it's a good time to upgrade that types and ec-ts at once. Will look into it.

**Comment by rwjblue** [comment]
I rerolled the yarn.lock in #814 (since floating deps tests worked), mind trying again?
